### ISSUE TITLE
VPNaaS - model changes

### ISSUE THREAD

**Issue by shuwens** [shwsun/haas#3]
To support vpnaas, we will need some modifications to the `model.py`. E.g. Vpn node will support generating virtual nic and populate the database.

For every physical node allocated to serve VPN services, i.e. every `project_make_vpn` api call, we will create a virtual nic. It falls into the way that we used to do for headnodes, but now we want to host the headnodes  at anyplaces, I assume we need to have a `headnode.cfg` kinda file to do this. 

tap, br-vlan

**Comment by zenhack** [comment]
I'm not sure I follow what you're saying here?

I'm pretty sure the thinking was to gut the headnodes entirely; once we have the vpn in place folks can just spin up a VM somewhere (perhaps a nearby nova instance) and connect via openvpn; HIL doesn't need to care.

Going off the proposal I just wrote up in #4, the table for vpns would have these fields:
- project, the project to which the vpn belongs
- network, the network it should be attached to
- Some kind of unique identifier, perhaps a uuid.
- A udp port number

From there, we can save the config file and keys to paths that are a function of the uuid, e.g. `/foo/bar/openvpn-${uuid}.cfg`

systemd can have service scripts parametrized over some value, so to manage the lifecycle of that vpn, we can do things like `systemctl start openvpn@${uuid}`.

**Comment by shuwens** [comment]
@zenhack you are absolutely right, and I already finished the systemd script part, see https://github.com/shwsun/hil/blob/vpnaas/scripts/vpn%40.service

uuid is good point, I will think about it.

**Comment by zenhack** [comment]
I might suggest not using `--daemon` and `--writepid`; using pidfiles introduces a race condition and I don't see any reason we can't just have systemd run it in the foreground.

It also may be worth using posix caps/otherwise reducing the priviledges the openvpn process gets. Below is the stock service script on my machine.

```
[Unit]
Description=OpenVPN connection to %I
After=syslog.target network.target network-online.target
Documentation=man:openvpn(8)

[Service]
PrivateTmp=true
Type=forking
ExecStart=/usr/bin/openvpn --cd /etc/openvpn --config %i.conf --daemon openvpn@%i --writepid /run/openvpn@%i.pid --status-version 2
PIDFile=/run/openvpn@%i.pid
CapabilityBoundingSet=CAP_IPC_LOCK CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_SETGID CAP_SETUID CAP_SYS_CHROOT CAP_DAC_READ_SEARCH
LimitNPROC=10
DeviceAllow=/dev/null rw
DeviceAllow=/dev/net/tun rw

[Install]
WantedBy=multi-user.target
```

**Comment by shuwens** [comment]
1 more problem brought up by @matsuuran : changes for `vpnnode_connect_network`, `project_connect_node`

**Comment by zenhack** [comment]
`vpnnode_connect_network`? Do you just mean `node_connect_network`?

I don't think either of those needs any changes.
